qiskit_research/utils/custom_passes/periodic_dynamical_decoupling.py

```
@@ -0,0 +1,452 @@
+        # slack = 992 dt - 4 x 160 dt = 352 dt
+        #
+        # unconstraind sequence: 44dt-X1-88dt-Y2-88dt-X3-88dt-Y4-44dt
+        # constraind sequence  : 32dt-X1-80dt-Y2-80dt-X3-80dt-Y4-32dt + extra slack 48 dt
+        #
+        # Now we evenly split extra slack into start and end of the sequence.
+        # The distributed slack should be multiple of 16.
+        # Start = +16, End += 32
+        #
+        # final sequence       : 48dt-X1-80dt-Y2-80dt-X3-80dt-Y4-64dt / in total 992 dt
+        #
+        # Now we verify t0 of every node starts from multiple of 16 dt.
+        #
+        # X1:  48 dt (3 x 16 dt)
+        # Y2:  48 dt + 160 dt + 80 dt = 288 dt (18 x 16 dt)
+        # Y3: 288 dt + 160 dt + 80 dt = 528 dt (33 x 16 dt)
+        # Y4: 368 dt + 160 dt + 80 dt = 768 dt (48 x 16 dt)
+        #
+        # As you can see, constraints on t0 are all satified without explicit scheduling.
+        time_interval = t_end - t_start
+
+        if self._qubits and dag.qubits.index(qubit) not in self._qubits:
+            # Target physical qubit is not the target of this DD sequence.
+            self._apply_scheduled_op(
+                dag, t_start, Delay(time_interval, dag.unit), qubit
+            )
+            return
+
+        if self._skip_reset_qubits and (
+            isinstance(prev_node, DAGInNode) or isinstance(prev_node.op, Reset)
+        ):
+            # Previous node is the start edge or reset, i.e. qubit is ground state.
+            self._apply_scheduled_op(
+                dag, t_start, Delay(time_interval, dag.unit), qubit
+            )
+            return
+
+        slack = time_interval - np.sum(self._base_dd_sequence_lengths[qubit])
+        sequence_gphase = self._sequence_phase
+
+        if slack <= 0:
+            # Interval too short.
+            self._apply_scheduled_op(
+                dag, t_start, Delay(time_interval, dag.unit), qubit
+            )
+            return
+
+        if len(self._base_dd_sequence) == 1:
```

we should throw an error in this case